frontend/src/metabase/visualizations/components/ChartSettings/ChartSettings.tsx

```diff
@@ -356,60 +361,142 @@ export const ChartSettings = ({
-      )}
       <ChartSettingsWidgetPopover
         anchor={popoverRef as HTMLElement}
         widgets={[styleWidget, formattingWidget].filter(
           (widget): widget is Widget => !!widget,
         )}
         handleEndShowWidget={handleEndShowWidget}
       />
+    </>
+  );
+};
+
+export const QuestionChartSettings = ({
+  question,
+  widgets: propWidgets,
+  series,
+  onChange,
+  computedSettings,
+  initial,
+}: QuestionChartSettingsProps) => {
+  const { chartSettings, handleChangeSettings, transformedSeries } =
+    useChartSettingsState({ series, onChange });
+
+  const widgets = useMemo(
+    () =>
+      propWidgets ||
+      getSettingsWidgetsForSeries(
+        transformedSeries,
+        handleChangeSettings,
+        false,
+      ),
+    [propWidgets, transformedSeries, handleChangeSettings],
+  );
+
+  return (
+    <ChartSettingsRoot>
+      <ChartSettings
+        question={question}
+        series={series}
+        onChange={onChange}
+        initial={initial}
+        computedSettings={computedSettings}
+        chartSettings={chartSettings}
+        transformedSeries={transformedSeries}
+        widgets={widgets}
+      />
     </ChartSettingsRoot>
   );
 };
 
-export const ChartSettingsWithState = (props: ChartSettingsWithStateProps) => {
-  const [tempSettings, setTempSettings] = useState(props.settings);
+export const DashboardChartSettings = ({
+  className,
+  dashboard,
+  dashcard,
+  isDashboard,
+  onChange,
+  series,
+  onClose,
+  widgets: propWidgets,
+}: DashboardChartSettingsProps) => {
+  const [tempSettings, setTempSettings] = useState<VisualizationSettings>();
+  const [warnings, setWarnings] = useState<string[]>();
+
+  const {
+    chartSettings,
+    handleChangeSettings,
+    chartSettingsRawSeries,
+    transformedSeries,
+  } = useChartSettingsState({
+    series,
+    settings: tempSettings,
+    onChange: setTempSettings,
+  });
+
+  const handleDone = useCallback(() => {
+    onChange?.(chartSettings ?? tempSettings);
+    onClose?.();
+  }, [chartSettings, onChange, onClose, tempSettings]);
 
-  useEffect(() => {
-    if (props.settings) {
-      setTempSettings(props.settings);
-    }
-  }, [props.settings]);
+  const handleResetSettings = useCallback(() => {
+    const originalCardSettings = dashcard?.card.visualization_settings;
+    const clickBehaviorSettings = getClickBehaviorSettings(chartSettings);
 
-  const onDone = (settings: VisualizationSettings) =>
-    props.onChange?.(settings ?? tempSettings);
+    onChange?.({
+      ...originalCardSettings,
+      ...clickBehaviorSettings,
+    });
+  }, [chartSettings, dashcard?.card.visualization_settings, onChange]);
+
+  const onResetToDefault =
+    // resetting virtual cards wipes the text and broke the UI (metabase#14644)
+    !_.isEqual(chartSettings, {}) && (chartSettings || {}).virtual_card == null
+      ? handleResetSettings
+      : null;
+
+  const handleCancel = useCallback(() => {
+    onClose?.();
+  }, [onClose]);
```

Comment:
I don't think we need to wrap `useCallback` in `onClose` here. Could we pass `onClose` directly?

Reply:
yeah we should definitely do that